### DISCUSSION TITLE
Difference of output between in API vs in CLI

### DISCUSSION THREAD

**Discussion by nedpals** [tree-sitter/tree-sitter#1288] [Q&A] [answered]
Hello everyone! I'm currently working on a language server for the [V language](https://github.com/vlang/vls) and we recently migrated to tree-sitter (Currently at 0.19.5 I believe) and the results have been positive so far. 

I have a question though regarding the output. I have noticed that the following source code: 
```v
module main

fn main() {
  two := 2
  mut zero := 0
  
  two = 
}
```
Produces the following syntax tree:
```
(MISSING ERROR 
     (module_clause (module_identifier)) 
     name: (identifier) (parameter_list) 
     (short_var_declaration 
          left: (identifier_list (identifier)) 
          right: (expression_list (int_literal))) 
     (short_var_declaration 
          left: (identifier_list (identifier)) 
          right: (expression_list (int_literal))) 
     (expression_list (identifier)))
```

Meanwhile in the CLI version, it produces this syntax tree: 
```
(source_file [0, 0] - [7, 1]
  (module_clause [0, 0] - [0, 11]
    (module_identifier [0, 7] - [0, 11]))
  (function_declaration [2, 0] - [7, 1]
    name: (identifier [2, 3] - [2, 7])
    parameters: (parameter_list [2, 7] - [2, 9])
    body: (block [2, 10] - [7, 1]
      (short_var_declaration [3, 2] - [3, 10]
        left: (identifier_list [3, 2] - [3, 5]
          (identifier [3, 2] - [3, 5]))
        right: (expression_list [3, 9] - [3, 10]
          (int_literal [3, 9] - [3, 10])))
      (short_var_declaration [4, 2] - [4, 15]
        left: (identifier_list [4, 2] - [4, 10]
          (identifier [4, 6] - [4, 10]))
        right: (expression_list [4, 14] - [4, 15]
          (int_literal [4, 14] - [4, 15])))
      (assignment_statement [6, 2] - [6, 7]
        left: (expression_list [6, 2] - [6, 5]
          (identifier [6, 2] - [6, 5]))
        right: (expression_list [6, 7] - [6, 7]
          (array [6, 7] - [6, 7]))))))
hello.v 57 ms   (MISSING "[]" [6, 7] - [6, 7])
```

The difference the former and the latter is that the latter was able to recover and build the `function_declaration` node while the former skips it. Is there anyway I can replicate the same tree output as the CLI's?

**Comment by razzeee** [comment]
Hrm, might be related https://github.com/tree-sitter/tree-sitter/issues/1211

Is this using WASM too? Do you have an external scanner involved?

**Reply by nedpals** [reply]
@Razzeee Not really, we use the C library directly with our own language bindings and it does not use an external scanner at all

**Reply by ahlinc** [reply]
@nedpals Could you get a link to your tree-sitter grammar for the V language?

**Reply by nedpals** [reply]
> @nedpals Could you get a link to your tree-sitter grammar for the V language?

Sure thing. Here's the link: https://github.com/nedpals/tree-sitter-v

**Reply by ahlinc** [reply]
@nedpals Thanks! Could you also describe how I can reproduce your first output that starts with `(MISSING ERROR ` ?

**Reply by nedpals** [reply]
> @nedpals Thanks! Could you also describe how I can reproduce your first output that starts with `(MISSING ERROR ` ?

The best way I can describe in a non-specific programming language fashion is that aside from setting the grammar language to V, I just basically parse the said source code with `ts_parser_parse_string` and print the S-expression with `ts_node_string`. Hope it helps :))

**Comment by nedpals** [comment, marked as answer]
This is weird. I tried using it directly in C and showed the appropriate results. Might be an issue to the to the input data type from my side. My bad ://

**Reply by nedpals** [reply]
Just for heads up, in case others might encounter this, here's what the investigation I did. I wrote a simple C program (which came from the [docs](https://tree-sitter.github.io/tree-sitter/using-parsers#an-example-program)) for testing:
```c
#include <stdio.h>
#include <string.h>
#include <tree_sitter/api.h>

TSLanguage *tree_sitter_v();

int main() {
  TSParser *parser = ts_parser_new();
  ts_parser_set_language(parser, tree_sitter_v());

  char *source_code = "module main\n\nfn main() {\n  two := 2\n  mut zero := 0\n  \n  two = \n}";
  TSTree *tree = ts_parser_parse_string(parser, NULL, source_code, strlen(source_code));

  TSNode root_node = ts_tree_root_node(tree);
  char *sexpr = ts_node_string(root_node);
  printf("Syntax tree: %s\n", sexpr);

  free(sexpr);
  ts_tree_delete(tree);
  ts_parser_delete(parser);
  return 0;
}
```
This was the output I received which was what I expected:
```
Syntax tree: (source_file (module_clause (module_identifier)) (function_declaration name: (identifier) parameters: (parameter_list) body: (block (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (assignment_statement left: (expression_list (identifier)) right: (expression_list (array (MISSING "[]")))))))
```

I then wrote the same program but in the language I was using (in this case, V):
```v
module main

import tree_sitter
import tree_sitter_v.bindings.v

fn main() {
    mut parser := tree_sitter.new_parser()
    parser.set_language(v.language)

    source_code := 'module main\n\nfn main() {\n  two := 2\n  mut zero := 0\n  \n  two = \n}'
    tree := C.ts_parser_parse_string(parser, 0, &char(source_code.str), source_code.len)

    root_node := C.ts_tree_root_node(tree)
    sexpr := unsafe { C.ts_node_string(root_node).vstring() }
    println('Syntax tree: $sexpr')
}
```
And ran it as well which caused the error I had faced in the issue:
```
Syntax tree: (MISSING ERROR (module_clause (module_identifier)) name: (identifier) (parameter_list) (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (expression_list (identifier)) (MISSING "_ERROR"))
```

Now I'm not sure if the C char pointer on V used for interacting with C functions has the same structure as C's, but the solution I end up with in the end was to subtract the string length passed to the `ts_parser_parse_string` function by 1:
```v
// ...
C.ts_parser_parse_string(parser, 0, &char(source_code.str), source_code.len - 1)
// ...
```

Which ended up with the current result that I'm getting right now. Now quite there yet but at least it improved the situation a bit:
```
Syntax tree: (MISSING source_file (module_clause (module_identifier)) (function_declaration name: (identifier) parameters: (parameter_list)) (ERROR (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (short_var_declaration left: (identifier_list (identifier)) right: (expression_list (int_literal))) (expression_list (identifier))))
```
